### PR TITLE
path-util: introduce several path utilities and handle "." more gracefully

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/basic/path-util.c [hunk 1]
```diff
@@ -794,6 +794,89 @@ char* dirname_malloc(const char *path) {
...
+                if (startswith(p, "./")) {
+                        p++;
+                        continue;
+                }
+                break;
+        }
+        return p;
+}
+
+int path_get_first_component(const char **p, bool accept_dot_dot, const char **ret) {
```
Comment (poettering) [inline src/basic/path-util.c]: maybe call this "path_find_first_component()"? 

i.e. "get" sounds you'd actually return something you actively acquire. but this here is just about finding the position of two concepts in the path

File: src/basic/path-util.c [hunk 2]
```diff
@@ -794,6 +794,89 @@ char* dirname_malloc(const char *path) {
         return dir2;
 }
 
+static const char *drop_slash_or_dot(const char *p) {
```
Comment (poettering) [inline src/basic/path-util.c]: maybe "skip_slash_or_dot()"?

**Comment by yuwata** [comment]
Now, I think this is ready for review. I added the last two commits after @poettering reviewed last time. So, PTAL!

**Comment by yuwata** [comment]
@poettering Could you take a look?

File: src/basic/path-util.c [hunk 3]
```diff
@@ -794,6 +794,89 @@ char* dirname_malloc(const char *path) {
...
+                        *ret = NULL;
+                return 0;
+        }
+
+        end_first = strchrnul(first, '/');
+        len = end_first - first;
+
+        if (len > NAME_MAX)
+                return -EINVAL;
+        if (!accept_dot_dot && len == 2 && strneq(first, "..", 2))
```
Comment (poettering) [inline src/basic/path-util.c]: i am pretty sure for such a simple check a `first[0] == '.' && first[1] == '.'` would be quicker and simpler

File: src/test/test-path-util.c [hunk 4]
```diff
@@ -550,6 +550,88 @@ static void test_file_in_same_dir(void) {
...
+        test_path_find_first_component_one("./", false, NULL, 0);
+        test_path_find_first_component_one("./.", false, NULL, 0);
+        test_path_find_first_component_one("..", false, NULL, -EINVAL);
+        test_path_find_first_component_one("/..", false, NULL, -EINVAL);
+        test_path_find_first_component_one("./..", false, NULL, -EINVAL);
+        test_path_find_first_component_one("////./././//.", false, NULL, 0);
+        test_path_find_first_component_one("a/b/c", false, STRV_MAKE("a", "b", "c"), 0);
+        test_path_find_first_component_one("././//.///aa/bbb//./ccc", false, STRV_MAKE("aa", "bbb", "ccc"), 0);
+        test_path_find_first_component_one("././//.///aa/.../../bbb//./ccc/.", false, STRV_MAKE("aa", "..."), -EINVAL);
+        test_path_find_first_component_one("//./aaa///.//./.bbb/..///c.//d.dd///..eeee/.", false, STRV_MAKE("aaa", ".bbb"), -EINVAL);
```
Comment (poettering) [inline src/test/test-path-util.c:655]: could you add some tests with a string like `a/foo./b` ,i.e. with a ./ at the end of a valid path component?

File: src/basic/path-util.c [hunk 5]
```diff
@@ -1023,28 +1023,21 @@ bool filename_is_valid(const char *p) {
...
 bool path_is_valid(const char *p) {
-
         if (isempty(p))
                 return false;
 
         for (const char *e = p;;) {
-                size_t n;
+                int r;
+
+                r = path_find_first_component(&e, true, NULL);
```
Comment (poettering) [inline src/basic/path-util.c]: maybe insert `/* accept_dot_dot= */` before the `true`?

File: src/basic/path-util.c [hunk 6]
```diff
@@ -794,6 +794,89 @@ char* dirname_malloc(const char *path) {
...
+}
+
+int path_find_first_component(const char **p, bool accept_dot_dot, const char **ret) {
+        const char *q, *first, *end_first, *next;
+        size_t len;
+
+        assert(p);
+
+        /* When a path is input, then returns the pointer to the first component and its length, and
+         * move the input pointer to the next component or nul.
```
Comment (poettering) [inline src/basic/path-util.c]: maybe also add a sentence:

```
This skips both over any `/` immediately *before* and *after* the first component before returning.
```

I think that's important to realize when using the function.

File: src/basic/path-util.c [hunk 7]
```diff
@@ -448,28 +448,25 @@ char* path_startswith(const char *path, const char *prefix) {
...
+                int r, s;
 
-                if (*prefix == 0)
-                        return (char*) path;
+                r = path_find_first_component(&path, accept_dot_dot, &p);
+                if (r < 0)
+                        return NULL;
 
-                if (*path == 0)
+                s = path_find_first_component(&prefix, accept_dot_dot, &q);
```
Comment (poettering) [inline src/basic/path-util.c]: doesn't matter, but in 95% of all cases local variables called `s` refers to a string in functions that primarily deal with strings... maybe rename `k`? 

but really doesn't matter

File: src/basic/hash-funcs.c [hunk 8]
```diff
@@ -27,30 +27,31 @@ void path_hash_func(const char *q, struct siphash *state) {
...
-                q += n;
+                const char *e;
+                int r;
+
+                r = path_find_first_component(&q, true, &e);
+                if (r == 0)
+                        return;
+
+                if (add_slash)
+                        siphash24_compress("/", 1, state);
```
Comment (poettering) [inline src/basic/hash-funcs.c]: could also be written a `siphash24_compress_byte('/', state);`

but doesn't matter at all

File: src/basic/fs-util.c [hunk 9]
```diff
@@ -1007,24 +974,20 @@ int chase_symlinks(const char *path, const char *original_root, unsigned flags,
...
-                                /* If done is "/", as first also contains slash at the head, then remove this redundant slash. */
-                                if (streq_ptr(done, "/"))
-                                        *done = '\0';
-
-                                if (!strextend(&done, first, todo))
+                                joined = path_join(done, first, todo);
+                                if (!joined)
                                         return -ENOMEM;
 
+                                free_and_replace(done, joined);
```
Comment (poettering) [inline src/basic/fs-util.c:990]: why not strextend()?

File: src/basic/fs-util.c [hunk 10]
```diff
@@ -1079,27 +1041,20 @@ int chase_symlinks(const char *path, const char *original_root, unsigned flags,
...
-                                joined = path_join("/", destination, todo);
+                        /* Prefix what's left to do with what we just read, and start the loop again, but
+                         * remain in the current directory. */
+                        joined = path_join(destination, todo);
                         if (!joined)
                                 return -ENOMEM;
 
-                        free(buffer);
-                        todo = buffer = joined;
+                        free_and_replace(buffer, joined);
```
Comment (poettering) [inline src/basic/fs-util.c:1056]: strextend()?

File: src/basic/fs-util.c [hunk 11]
```diff
@@ -1108,29 +1063,16 @@ int chase_symlinks(const char *path, const char *original_root, unsigned flags,
...
-                        if (streq(done, "/"))
-                                *done = '\0';
-
-                        if (!strextend(&done, first))
-                                return -ENOMEM;
-                }
+                joined = path_join(done, first);
+                if (!joined)
+                        return -ENOMEM;
+                free_and_replace(done, joined);
```
Comment (poettering) [inline src/basic/fs-util.c]: one of those days we should add path_extend(), or so ;-)

**Comment by yuwata** [comment]
@poettering Thank you for your review. All suggestions are applied. Also, added a commit at the beginning which fixes an off by one issue in `path_extend()`.

Upgrading the green label.

**Pull request merged**